homeassistant/components/sensor/__init__.py

```diff
@@ -570,12 +570,18 @@ def state(self) -> Any:
 
         numerical_value: int | float | Decimal | None = None
         numerical_conversion_failed = None
-        if not isinstance(value, (int, float, Decimal)):
-            try:
-                numerical_value = float(value)  # type:ignore[arg-type]
-            except (TypeError, ValueError):
-                numerical_conversion_failed = True
-        else:
+        if isinstance(value, str):
+            if "." in value:
+                try:
+                    numerical_value = float(value)
+                except (TypeError, ValueError):
```

Comment:
TypeError cannot occur after `isinstance(value, str)`
The reason for `TypeError` was in case an invalid value is passed by a sensor (eg. a date/datetime without the device class set)

That was also the reasoning behind `# type:ignore[arg-type]`

I'm not sure what the "optimum" way to resolve this is...

Maybe
```python
if not isinstance(value, (int, float, Decimal)):
            try:
                if isinstance(value, str) and "." in value:
                    numerical_value = int(value)  # type:ignore[arg-type]
                else:
                    numerical_value = float(value)  # type:ignore[arg-type]
            except (TypeError, ValueError):
                numerical_conversion_failed = True
        else:
            numerical_value = value
```

Reply:
Shouldn't we only try to convert to a number if the value is a string though, in that case we can just remove catching the `TypeError`?

Reply:
Possibly - that's what I'm not 100% sure about.
Are there other types (appart from int/float/Decimal/str) that could be converted to int or float?
For other types that would have raised a TypeError, we should then at least set `numerical_conversion_failed = True`